criu/files-reg.c

```diff
@@ -812,6 +843,36 @@ static int dump_ghost_file(int _fd, u32 id, const struct stat *st, dev_t phys_de
 		gfe.size = st->st_size;
 	}
 
+	/*
+	 * We set gfe.symlnk_target only if we need to dump
+	 * symlink content, otherwise we leave it NULL.
+	 * It will be taken into account on restore in mklnk_ghost function.
+	 */
+	if (S_ISLNK(st->st_mode)) {
+		int ret;
+
+		/*
+		 * We assume that _fd opened with O_PATH | O_NOFOLLOW
+		 * flags because S_ISLNK(st->st_mode). With current kernel version,
+		 * it's looks like correct assumption in any case.
+		 */
+		ret = readlinkat(_fd, "", pathbuf, sizeof(pathbuf) - 1);
+		if (ret < 0) {
+			pr_perror("Can't readlinkat");
+			goto err_out;
+		}
+
+		pathbuf[ret] = 0;
+
+		if (ret != st->st_size) {
+			pr_err("Buffer for readlinkat is too small: ret %u, st_size %u, buf %u %s\n",
+					(unsigned)ret, (unsigned)st->st_size, (unsigned)PATH_MAX, pathbuf);
```

Comment:
`ret %d` (or better make it ssize_t to be the same type as syscall)
`size %"PRIu64" `
and no need for any cast..

Reply:
> `ret %d` (or better make it ssize_t to be the same type as syscall)
> `size %"PRIu64" `
> and no need for any cast..

I'm sorry, PRId64 I think, as ssize_t is signed type. Right?

Reply:
Yep :)